pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp

```diff
@@ -71,232 +104,412 @@ class PilzJointTrajectoryControllerTest : public testing::Test
+/**
+ * @tests{end_holding,
+ * Tests unholding the controller at different time points after the controller is started.
+ * }
+ */
+TEST_F(PilzJointTrajectoryControllerTest, testUnholdSuccess)
+{
+  ASSERT_TRUE(manager_->loadController()) << "Failed to initialize the controller.";
+  ASSERT_TRUE(waitForActionServer());
 
-  trajectory_action_client_.sendGoal(goal);
+  manager_->startController();
 
-  ros::Duration period{DEFAULT_UPDATE_PERIOD_SEC};
-  progressInTime(period);
-  updateController();
+  std_srvs::TriggerRequest req;
+  std_srvs::TriggerResponse resp;
+  EXPECT_TRUE(manager_->triggerUnHold(req, resp));
+  EXPECT_FALSE(resp.success);
 
-  trajectory_action_client_.waitForResult();
-  EXPECT_EQ(control_msgs::FollowJointTrajectoryResult::INVALID_GOAL, trajectory_action_client_.getResult()->error_code)
-      << "Obtained error code " << trajectory_action_client_.getResult()->error_code << " for goal while in HOLD mode.";
-}
+  ros::Duration stop_duration{STOP_TRAJECTORY_DURATION_SEC + 2*DEFAULT_UPDATE_PERIOD_SEC};
+  progressInTime(stop_duration);
+  update();
 
+  EXPECT_TRUE(manager_->triggerUnHold(req, resp));
+  EXPECT_TRUE(resp.success);
+  EXPECT_TRUE(isControllerInUnholdMode());
+}
 
 /**
- * @brief Test for correct return values and elapsed times in case of subsequent calls to hold/unhold function.
- *
- * Test Sequence:
- *    0. Start controller and perform initial update.
- *    1. Request DEFAULT twice.
- *    2. Request HOLDING.
- *    3. Request HOLDING again.
- *
- * Expected Results:
- *    0. Controller is executing (entering HOLD mode at startup).
- *    1. Each Response contains success==True.
- *    2. Response contains success==True. The elapsed time corresponds to the expected stop trajectory duration.
- *    3. Response contains success==True. The second call does not trigger a stop trajectory.
+ * @tests{start_holding,
+ * Tests holding the controller.
+ * }
+ * @tests{no_execution_during_hold,
+ * Tests holding the controller.
+ * }
  */
-TEST_F(PilzJointTrajectoryControllerTest, testDoubleRequest)
+TEST_F(PilzJointTrajectoryControllerTest, testHoldSuccessAfterUnhold)
 {
-  ASSERT_TRUE(controller_->init(hardware_, nh_, controller_nh_)) << "Failed to initialize the controller.";
-  controller_->state_ = controller_->INITIALIZED;
-  startController();
+  ASSERT_TRUE(performFullControllerStartup());
+
+  std_srvs::TriggerRequest req;
+  std_srvs::TriggerResponse resp;
+  // run async such that stop trajectory can be executed in the meantime
+  std::future<bool> hold_future = manager_->triggerHoldAsync(req, resp);
+
+  std::chrono::milliseconds hold_timeout {WAIT_FOR_HOLD_FUTURE_MSEC};
+  EXPECT_TRUE(waitFor([&hold_future](){ return isFutureReady(hold_future); }, hold_timeout, true));
+  EXPECT_TRUE(resp.success);
+  EXPECT_TRUE(isControllerInHoldMode());
+}
 
-  controller_->state_ = controller_->RUNNING;
+/**
+ * @tests{start_holding,
+ * Tests holding the controller.
```

Comment:
You changed the docu below but not this one. Please update the docu, too.

Reply:
Thanks!